### ISSUE TITLE
Conversion to Pandas DataFrame fails when involving empty branches

### ISSUE THREAD

**Issue by chernals** [scikit-hep/uproot#396]
Not sure that the title correctly describes the issue.

Basically I have this:

`ValueError: cannot use flatten=True on branches with different jagged structure, such as electrons and muons (different, variable number of each per event); either explicitly select compatible branches, such as ["MET_*", "Muon_*"] (scalar and variable per event is okay), or set flatten=False`

I understand that and it makes sense. However, in my case, it fails because one branch (actually a "sub-branch", ie. a branch with a `splitlevel > 0`) is empty. I expect the branch to have the correct structure if filled at all.

I imagine it would be possible to handle that (corner) case smoothly, so that I can have the same code wether or not the branch is filled. If empty, it could be indicated by something with `pd.isnull(x) == True` for all subentries.

**Comment by jpivarski** [comment]
All of the branches passed into a DataFrame must have the same number of entries, but the TTree structure usually produces branches of exactly the same number of entries (`numentries`), anyway.

If one branch is jagged and another is not, then everything is still fine. If two branches are jagged, they must have exactly the same number of subentries in each entry.

So I'm confused—if one branch is empty (zero entries? zero subentries?) then it can't be combined with another that is not. Can you clarify your exact situation? (Supplying a small sample file works, but for this, it might be easy enough to give me some types and numbers of entries and subentries.)

**Comment by chernals** [comment]
Let's imagine something like this. I have a branch "Foo":

`class Foo: public TObject
{
public:
  int                       n = 0;   ///< Number of entries
  std::vector<float>        energy;  ///< Energy deposited in step
  std::vector<int>          partID; 
}`

which will process hits occuring during an event, using the branch "fill" method:

`
void Foo::Fill(const Bar* hit)
{
  n++;
  energy.push_back(hit->GetEnergy() );

  if (my_condition)
    {
      partID.push_back(hit->GetPartID() );
   }
`

So basically, in order to save some storage space, I might not store / populate the "partID" vector. This will result in a branch with jagged arrays of unequal sizes, because one of them will be of size zero.

However, I would like to have a dataframe will all 4 columns in all cases.

I hope this clarifies what I mean, if not, I'll get a small size output file and some Python code to illustrate.

**Comment by jpivarski** [comment]
So, unlike "jets" and "muons," which are two unrelated sets (unrelated apart from being in the same event and maybe having common physics origins, but not necessarily one-to-one), there's a natural mapping from each "partID" to "energy," though it's not a bijection because some "partIDs" are missing.

However, there's not enough information here to determine what that mapping is: if there are 3 "energies" and 2 "partIDs," which one is missing—the first, the second, or the third? We'd like to be able to fill a DataFrame in which the "partID" column is `[n/a, #, #]`, `[#, n/a, #]`, or `[#, #, n/a]`, to align with the "energy" column (`[#, #, #]`), but we can't. Perhaps the booleans for `my_condition` could be passed through, but a general library like uproot won't know which column to use as the column—this is a domain-specific thing that an analysis script has to match up manually.

If you want these data to be turned into Pandas DataFrames, you should pass through `NaN` if `my_condition` is false, maintaining equal lengths for "energy" and "partID". If "partID" is floating point, you can fill them with `sqrt(-1)`. If it's an integer type, you'll have to pick a dummy value like `-1` and convert the dummy value to `NaN` in Pandas. (Pandas can assign `NaN` to integer types, but most other systems can't.)

Re-open if I've misunderstood. Thanks!

**Comment by chernals** [comment]
The part that I didn't explain clearly is that `my_condition` is an external condition, set up for all events, all the time while the (simulation) data are being dumped to root.

Basically what I'm suggesting is to broadcast zero-length jagged array onto a jagged array of nan's of the maximum length for that event (based on the other lengths, which are all equal).

**Comment by jpivarski** [comment]
This is not something uproot can guess, so I don't see any appropriate changes to its automatic behavior.

It sounds to me, though, like you need to do some processing on the jagged arrays before filling the DataFrame. You can use `array.pad(length)` to make the length of each subarray within `array` to be at least `length` by padding with `None`. If you don't like the `MaskedArrays` that result, you can call `array.pad(length).fillna(-1)` to turn all of the `None` values into `-1`, for instance.

But putting `None` _after_ all the "partID" values could be wrong if `my_condition` doesn't have all `true` values before the first `false` (e.g. if they're independently random). If you have `my_condition` as an additional array, you can do it more intelligently to get the `None` values in the right places.

If you want uproot to go directly to Pandas because it does the flattening into an entry/subentry MultiIndex, then be aware that you can do

```python
df = awkward.topandas(jaggedarray, flatten=True)
```

to do the same thing (as a newer implementation that can go arbitrarily deep in jaggedness).
